typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-hetero.rkt

```diff
@@ -90,6 +108,11 @@
     (match (single-value #'vec)
       [(tc-result1: (and vec-t (app resolve (Is-a: (HeterogeneousVector: es)))))
        (tc/hetero-ref #'index es vec-t "vector")]
+      [(tc-result1: (and vec-t (app resolve (Union: _ ts)))) (=> continue)
+       (define es (hetero-vecs->elems ts))
```

Comment:
`(tc-result1: (and vec-t (app resolve (Union: _ (app hetero-vecs->elems (? list? es)))))`

would get rid of the need for the `(=> continue)`